### ISSUE TITLE
slow in Cordova plugin

### ISSUE THREAD

**Issue by jyang** [jxcore/jxcore#399]
Hi,

I was excited at seeing the Express sample run using a Cordova plugin on Android but was disappointed at its performance. I was only able to reach 60 requests/second on a Nexus Player (and lower on a Nexus 5). Node.js on Raspberry Pi 2 can do ~300. I looked a little and found mozjs and not v8 in the plugin.

Is this performance about right? Is there a way to use V8 on Android in the Cordova plugin?

BTW, I didn't find a discussion group for JXcore and thus this bug.

Thanks!

**Comment by obastemur** [comment]
60 request per second is engine independent and expected for the Android side. For now the Android side of this project uses UI thread temporarily. 

Luckily JXcore can create sub threads.

Try this;
```
function Task() {
  // put your server code here
}

jxcore.tasks.addTask(Task);
```
and test it again.

**Comment by jyang** [comment]
I wrapped the Express code in a task:

```
jxcore.tasks.addTask(function() {
  clog('>>> task running');
  var express = require('express');
  var app = express();

  app.get('/', function (req, res) {
    res.send('Hello World! (' + Date.now() + ")");
    clog("Request", req.headers['x-forwarded-for'] ||
        req.connection.remoteAddress ||
        req.socket.remoteAddress ||
        req.connection.socket.remoteAddress, new Date());
  });

  var server = app.listen(3000, function () {
    clog("Express server is started. (port: 3000)");
  });
});
clog('>>> task added');
```

But I only see ">>> task added" and never ">>> task running". How to get the task to run?

**Comment by obastemur** [comment]
@jyang I've added an entirely new express sample into cordova plugin that uses a subthread for express server. Sync to latest and use that sample please.

**Comment by ktrzeciaknubisa** [comment]
@jyang Also, from what I see in your code, when you are calling inside a task `clog('>>> task running');` - the `clog` is not defined here (probably in a main thread). Threads are separate instances and cannot use functions/objects defined in other threads.

It is also mentioned in the sample which @obastemur talks about.

**Comment by jyang** [comment]
Thanks for the change. I tested it on a Nexus 5 and a Nexus Player. On Nexus 5, the highest throughput I got was 117 requests/second. Most of the time it was around 80-90. On Nexus Player, it was similar. This is roughly a 100% improvement in the average case, but still quite far from Pi 2's 300.

Also I noticed response times are widely distributed, ranging from ~50ms to over 1s, most of the time in the range of 100-300ms, but quite different from one request to the next.

Any further tips on squeezing more out of JXcore? Thanks!

**Comment by obastemur** [comment]
`~50ms to over 1s,`

Hmm. This sounds like the main UI looper timeout.  (BTW, did you remove clogs before benchmark?) and probably you make your load test over ethernet on RPI however use Wifi for mobile right?

**Comment by jyang** [comment]
Yes, I did remove the clog in server code. Load test was done with Pi on wired and Android over Wifi (with laptop that sends the load as the access point).

Please explain the UI loop. Isn't the server code already on a different thread?

**Comment by obastemur** [comment]
Wifi vs cable makes a huge difference in terms of multiple connections and lag

**Comment by plika** [comment]
You might create a network on the phone over USB tethering.
